### PR TITLE
Continuous testing

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This adds continuous testing support to Quarkus.

### CODE REVIEW

**Comment by geoand** [comment]
Cool! I'll start looking into the magic!

File: build-parent/pom.xml
Comment (gsmet) [inline build-parent/pom.xml:502]: We shouldn't do that, we should override the rule for the artifacts needing it. Otherwise we will definitely get it wrong (we added this rule for a good reason).

I understand it wasn't a priority until now but let's not merge it this way. Maybe @famod can help.

File: core/deployment/src/main/java/io/quarkus/deployment/TestConfig.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/TestConfig.java:26]: Looks like the javadoc is a bit off now that's it's an enum.

File: core/deployment/src/main/java/io/quarkus/deployment/TestConfig.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/TestConfig.java:28]: Better use lowercase paused as the accepted values will be lowercase in the documentation.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a couple of comments on the build and config, nothing major. I'll try to test it soon.

**Comment by gsmet** [comment]
I need the CI bandwitdh, I'll restart as soon as I have the 1.13.2 backport PR passing.

File: build-parent/pom.xml
Comment (famod) [reply build-parent/pom.xml:502]: I can have a look at the branch later this week (or more or less shortly after it was merged).

Comment (stuartwdouglas) [reply build-parent/pom.xml:502]: I updated it, I think it should be ok now.

**Comment by gsmet** [comment]
Rebased to get the latest CI fixes and retrigger CI.

File: build-parent/pom.xml
Comment (famod) [reply build-parent/pom.xml:492]: So you basically don't check transitively for junit anymore, right?
I wonder if we should narrow it down even more by only overriding `searchTransitive` to false in _specific_ modules?

Comment (stuartwdouglas) [reply build-parent/pom.xml:492]: Because testing is implemented in core/deployment every -deployment artifact would need it.

**Comment by gsmet** [comment]
I tested it a bit. It works great AFAICS and it's really useful. And I like the fact that it's disabled by default but easy to enable directly from the command line.

How does it work when you have longer tests though? It stops the current run when you save something and start a new one?

It still needs some UI love but I prefer we fine tune things once this is in. I'm pretty sure @phillip-kruger will be able to help on this too.

Nice work!

**Comment by phillip-kruger** [comment]
I would love to help !!! I'll look at this a.s.a.p :)

**Comment by gsmet** [comment]
@stuartwdouglas when I stopped the dev mode, I had this error:
```
2021-04-15 15:44:56,591 INFO  [io.quarkus] (Shutdown thread) validation-quickstart stopped in 0.025s
java.lang.IllegalStateException: Shutdown in progress
	at java.base/java.lang.ApplicationShutdownHooks.remove(ApplicationShutdownHooks.java:82)
	at java.base/java.lang.Runtime.removeShutdownHook(Runtime.java:244)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.close(IsolatedDevModeMain.java:313)
	at io.quarkus.deployment.dev.IsolatedDevModeMain$5.run(IsolatedDevModeMain.java:419)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
This was with a very simple quickstart (the validation one), nothing fancy.

**Comment by stuartwdouglas** [comment]
I noticed that late yesterday, its a simple fix.

File: devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
Comment (stuartwdouglas) [inline devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:164]: Note that this now means that quarkusDev will fail in gradle if the tests don't compile. In maven I can easily just ignore the failures, but I am not sure how to do that in Gradle.

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (stuartwdouglas) [inline extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:22]: @cescoffier I think this needs to be addressed in Mutiny. When you call setDefaultExecutor it gets wrapped in a ScheduledExecutor that starts a thread, which then can't really be shutdown without a hack like this.

This thread then ends up holding onto class loaders which is a big problem if you add too many dev mode tests to a module.

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java
Comment (stuartwdouglas) [inline extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: @mkouba I found a few places like this, name() already has the enclosing name (`EnclosingName$InnerName`) so this was ending up as `EnclosingName_EnclosingName$InnerName`, which then confused the application class predicate.

File: extensions/resteasy-classic/resteasy-jaxb/deployment/src/test/java/io/quarkus/resteasy/jaxb/deployment/ConsumesXMLTestCase.java
Comment (stuartwdouglas) [inline extensions/resteasy-classic/resteasy-jaxb/deployment/src/test/java/io/quarkus/resteasy/jaxb/deployment/ConsumesXMLTestCase.java:28]: No idea how this ever worked, the body() call does the serialization, so you need to set the content type first.

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java
Comment (mkouba) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: I'm pretty sure we used this for a reason so either something changed in the Jandex behavior or there's a corner case we tried to solve. Let me take a look...

Comment (mkouba) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: Ok, so I think that the correct version should be:
```java
if (bundleInterface.enclosingClass() != null) {
   baseName = DotNames.simpleName(bundleInterface.enclosingClass()) + "_"
          + DotNames.simpleName(bundleInterface);
} else {
   baseName = DotNames.simpleName(bundleInterface);
}
```
`DotNames.simpleName(bundleInterface.name())` -> `DotNames.simpleName(bundleInterface)`

And the resulting name should be `EnclosingName_InnerName` (at least that was the intent and we use this convention in ArC and some other extensions). The `DotNames.simpleName(DotName)` is not entirely correct because the dollar sign is a valid character for class names so we cannot detect a nested class here. 

> which then confused the application class predicate.

Do you remember the details? I'd like to verify this...

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (cescoffier) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:22]: We can add a method to shutdown the executor. 

Otherwise, I don't see how it can be fixed. Mutiny needs a Scheduled Executor (for several operators). We can delay the creation of the thread until the first call, however, even with this, once used, it will be there. 

CC @jponge

**Comment by stuartwdouglas** [comment]
As part of this PR I have had to change the test ClassLoading model to be the same as dev mode, to allow the test classes to be hot reloaded. This has shown up some problems (that would also be problems with devmode). I am hopeful that there is only a couple of these to go and then CI will be green.

File: extensions/mutiny/deployment/src/main/java/io/quarkus/mutiny/deployment/MutinyProcessor.java
Comment (stuartwdouglas) [inline extensions/mutiny/deployment/src/main/java/io/quarkus/mutiny/deployment/MutinyProcessor.java:16]: @cescoffier I have added this as a dep to vertx-http, so the correct executor is always setup, but given that this basically just sets up some infrastructure IMHO this extension is basically just an implementation detail, so should not be listed as a feature, WDYT?

Without adding this as a dep the dev mode tests leak all their ClassLoaders, as the default executors created by Mutiny prevent the CL from being freed.

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java
Comment (mkouba) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: I will address this in a separate PR. Pls revert this change.

Comment (mkouba) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: FTR https://github.com/quarkusio/quarkus/pull/16583

File: extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java
Comment (stuartwdouglas) [reply extensions/mutiny/runtime/src/main/java/io/quarkus/mutiny/runtime/MutinyInfrastructure.java:22]: You could also shut down the scheduled executor when a new one is being set.

From a Quarkus PoV what I have done here will stop the leaks, it just feels like a hack :-)

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java
Comment (stuartwdouglas) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: Does that actually work with the application class predicate though?

I am thinking now that I made a big mistake where I went for a flatish classpath in tests and the multiple ClassLoader's in dev mode, as things that work fine in tests and prod would actually fail in dev mode. 

Also this is part of a fairly massive changeset that bypasses all the incremental build stuff as it touches everything and really needs to be in by the alpha, so I am going to leave it as is and if this ever manages to pass CI and get in your change can go on top of it :-)

Comment (mkouba) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:524]: > Does that actually work with the application class predicate though?

It should work with the predicate used in the `MessageBundleProcessor`.

> Also this is part of a fairly massive changeset that bypasses all the incremental build stuff as it touches everything and really needs to be in by the alpha, so I am going to leave it as is and if this ever manages to pass CI and get in your change can go on top of it :-)

Well, I'd like to merge that PR and then there will be likely conflicts... so we'll see how it goes ;-).

**Comment by gsmet** [comment]
@stuartwdouglas the test failures look suspicious. I don't think I have seen these tests failing lately.

File: extensions/mutiny/deployment/src/main/java/io/quarkus/mutiny/deployment/MutinyProcessor.java
Comment (cescoffier) [reply extensions/mutiny/deployment/src/main/java/io/quarkus/mutiny/deployment/MutinyProcessor.java:16]: Not being listed is not a problem. With Resteasy Reactive always bringing it, it's often there (the fact it was listed was important when it was something to be added, and so easier for me to detect the missing feature).

About the dependency, no problem for me. 
In addition, @mkouba is looking at reorganizing the Vert.x extensions and would make the dependency directly into the vert.x core extension (and so expose the Mutiny variant of vert.x every time). The idea is to merge the vert.x extension and the vert.x core extension (mutiny is the only "extra Vert.x"  we support in 2.x, and merge the `@RouteFilter` into the `vertx-http` extension.

**Comment by stuartwdouglas** [comment]
It looks like an issue with Gradle testFixtures. I hopefully have a fix, testing it now. Not sure how it passed in Linux, and it passed on windows previously, so maybe these tests don't always fail properly?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/ClassScanResult.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/ClassScanResult.java:30]:
```suggestion
        ret.addedClasses.addAll(m1.addedClasses);
```

File: core/deployment/src/main/java/io/quarkus/deployment/dev/ClassScanResult.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/ClassScanResult.java:36]:
```suggestion
        ret.addedClasses.addAll(m2.addedClasses);
```

File: core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusCompiler.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusCompiler.java:158]: This condition would always be false, otherwise the line above would result in an NPE.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:198]: This could be fine but just to clarify it is. We would never have a `compileProblem` we wanted to report unless either the `configFileRestartNeeded` or the `ClassScanResult.isChanged()`?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:544]: Is `cuf` being applied twice to the same `module` on purpose (line 539 and 544)?

Comment (aloubyansky) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:544]: Looks like it's not a big deal.

**Comment by stuartwdouglas** [comment]
I have addressed the comments by @aloubyansky , it was green beforehand and the changes were minor so hopefully it stays that way.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java:115]: It doesn't look like this deployment CL is closed, is it? May lead to jars not being closed and locking issues on windows.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java:313]: Should this be wrapped in try-catch to make sure all the listeners are notified and the code after the loop gets executed?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestRunner.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestRunner.java:315]: A minor one, you could break the loop once the `application.properties` has been found.

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java:125]: Not necessarily for this PR, I am wondering whether we could pass the `AppModel` instance to the test app instead of re-resolving it. That could save some time.

File: devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
Comment (aloubyansky) [reply devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:164]: fyi @glefloch

File: devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java
Comment (aloubyansky) [inline devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:357]: fyi @glefloch

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/tests/TestsProcessor.java
Comment (aloubyansky) [inline extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/tests/TestsProcessor.java:107]: Looks like this `if` is present in 5 methods in this class, should it be extracted to something like `isDevConsoleEnabled(LaunchModeBuildItem)` or something?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java
Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java:125]: This is a TEST appmodel, but the outer one is DEV. I want to talk to you about this though, there is a lot of stuff that could be cleaned up here.

Comment (aloubyansky) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/testing/TestSupport.java:125]: Right, makes sense.

**Comment by stuartwdouglas** [comment]
I think I have figured out the intermittent Gradle failure. For some reason the test in question has the gradle jacoco plugin in the generated project, which is not used anywhere else. As we now resolve test dependencies this would now attempt to download this artifact, which may not always happen in 1 minute.

**Comment by stuartwdouglas** [comment]
@gsmet all green now

**Review by gsmet — APPROVED** [review]
OK, let's merge and iterate from there. This is a very significant and useful addition so I want to get this one in Alpha1 to get more eyeballs on it.

@stuartwdouglas I'll let you send me an email with a short description of the feature? Thanks!

**Pull request merged**
